Add Esm

pass-lin opened the pull request:
from https://github.com/keras-team/keras-hub/issues/2177
Achieved a smaller error with hf.
```python
import os
os.environ["KERAS_BACKEND"] = "torch"
os.environ["HF_ENDPOINT"] = "https://hf-mirror.com"

from keras import ops
from transformers.models.esm.modeling_esm import EsmAttention as hf_EsmSelfAttention
from transformers import EsmConfig
from esm2.esm2_layers import EsmSelfAttention
import numpy as np
import keras
from transformers.models.esm.modeling_esm import EsmModel
weights_path = "facebook/esm2_t6_8M_UR50D"
hf_model = EsmModel.from_pretrained(weights_path)
hf_model.cuda().eval()
hf_model.embeddings.token_dropout = False


from keras_hub.src.models.esm.esm_backbone import (
    ESMBackbone,
)


keras_model =  ESMBackbone.from_preset('hf://'+weights_path)
keras_model.summary()


x = ops.array([[1,2,3,4,5]])+1
hf_out = hf_model(x,ops.ones_like(x))[0]
keras_out = keras_model({'token_ids': x})

print(ops.all(ops.isclose(hf_out, keras_out,atol=1e-4)))



```
ESM Checkpoint Conversion and Numerics Verification Demo (across multiple backends): [Notebook Link](https://colab.research.google.com/drive/1IPnkQwxZ16E-CfygknU-z8-sCkitgo4l?usp=sharing)

Train Demo: [Notebook Link](https://colab.research.google.com/drive/1ALXd1Bq0sWns7ytWzNbwWNanh0YkzBY4?usp=sharing)

pass-lin commented:
```python
ruff.....................................................................Passed
ruff-format..............................................................Passed
Error: Process completed with exit code 1.
```
Please help me figure out how to solve this problem.

mattdangerw commented:
Probably an issue with generating the API symbols. Looks like you need to sync with the latest changes on master, then you could try running `./shell/api_gen.sh`

sachinprasadhs commented:
> ```python
> ruff.....................................................................Passed
> ruff-format..............................................................Passed
> Error: Process completed with exit code 1.
> ```
> 
> Please help me figure out how to solve this problem.

You can rebase it to latest master code
and then run -  `pre-commit run --all-files`
`pip install -u namex`

pass-lin commented:
```python
keras_hub/src/layers/modeling/reversible_embedding_test.py::ReversibleEmbeddingTest::test_quantize_dtype_argument_tie_weights - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/layers/modeling/reversible_embedding_test.py::ReversibleEmbeddingTest::test_quantize_dtype_argument_untie_weights - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/layers/modeling/reversible_embedding_test.py::ReversibleEmbeddingTest::test_quantize_int8_tie_weights - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/layers/modeling/reversible_embedding_test.py::ReversibleEmbeddingTest::test_quantize_int8_untie_weights - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/albert/albert_backbone_test.py::AlbertBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/bart/bart_backbone_test.py::BartBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/bert/bert_backbone_test.py::BertBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/bloom/bloom_backbone_test.py::BloomBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/clip/clip_backbone_test.py::CLIPBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/deberta_v3/deberta_v3_backbone_test.py::DebertaV3BackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/distil_bert/distil_bert_backbone_test.py::DistilBertBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/electra/electra_backbone_test.py::ElectraBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/f_net/f_net_backbone_test.py::FNetBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/falcon/falcon_backbone_test.py::FalconBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/gemma/gemma_backbone_test.py::GemmaBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/gemma/gemma_backbone_test.py::Gemma2BackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/gpt2/gpt2_backbone_test.py::GPT2BackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/gpt_neo_x/gpt_neo_x_backbone_test.py::GPTNeoXBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/llama/llama_backbone_test.py::LlamaTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/mistral/mistral_backbone_test.py::MistralBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/opt/opt_backbone_test.py::OPTBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/pali_gemma/pali_gemma_backbone_test.py::PaliGemmaBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/pali_gemma/pali_gemma_backbone_test.py::PaliGemma2BackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/phi3/phi3_backbone_test.py::Phi3Test::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/phi3/phi3_backbone_test.py::Phi3Test::test_backbone_basics_with_su_rotary - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/roberta/roberta_backbone_test.py::RobertaBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/siglip/siglip_backbone_test.py::SigLIPBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/siglip/siglip_backbone_test.py::SigLIP2BackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/t5/t5_backbone_test.py::T5BackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/whisper/whisper_backbone_test.py::WhisperBackboneTest::test_backbone_basics - TypeError: _int8_build() takes 2 positional arguments but 3 were given
FAILED keras_hub/src/models/xlm_roberta/xlm_roberta_backbone_test.py
```
@mattdangerw @sachinprasadhs 
Is it a problem with the test environment? Why are there so many errors that don't belong to me?

sachinprasadhs commented:
It's not related to your code, looks like some issue with the JAX backend, we will look into it.

sachinprasadhs commented on keras_hub/src/models/esm/esm_masked_plm.py:
```diff
@@ -0,0 +1,121 @@
+    training and evaluation. This is done by default when creating the model
+    with `from_preset()`.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind. The underlying model is provided by a
+    third party and subject to a separate license, available
+    [here](https://github.com/facebookresearch/esm).
+
```
Remove this, this will be part of the license and in the model card.

sachinprasadhs commented on keras_hub/src/models/esm/esm_classifier.py:
```diff
@@ -0,0 +1,119 @@
+    `fit()`, `predict()`, and `evaluate()`. This is done by default when
+    creating the model with `from_preset()`.
+
+    Disclaimer: Pre-trained models are provided on an "as is" basis, without
+    warranties or conditions of any kind.
+
```
Remove this

sachinprasadhs commented on keras_hub/src/utils/transformers/convert_esm.py:
```diff
@@ -0,0 +1,165 @@
+        "pad_token_id": transformers_config["pad_token_id"],
+        "max_sequence_length": transformers_config.get(
+            "max_position_embeddings", None
+        ),  # 默认值为None
+        "layer_norm_eps": transformers_config.get(
+            "layer_norm_eps", 1e-12
+        ),  # 默认值为1e-12
+        "emb_layer_norm_before": transformers_config.get(
+            "emb_layer_norm_before", False
+        ),  # 默认值为False
+        "activation": transformers_config.get(
+            "activation", "gelu"
+        ),  # 默认值为"gelu"
+        "max_wavelength": transformers_config.get(
+            "max_wavelength", 10000
+        ),  # 默认值为10000
+    }
```
Remove comments

sachinprasadhs reviewed (commented):
Thanks fro the PR, I have added my comments, also add checkpoints conversion under: keras-hub/tools/checkpoint_conversion

pass-lin commented:
@sachinprasadhs hey,please check update code

pass-lin commented:
@mattdangerw @sachinprasadhs 
Please check my code, thank you.

sachinprasadhs reviewed (commented):
Added few more comments and few of the previous review comments still needs to be addressed

sachinprasadhs commented:
Once you address all the comments, add end to end working colab along with the checkpoints conversion under: keras-hub/tools/checkpoint_conversion

pass-lin commented:
> Once you address all the comments, add end to end working colab along with the checkpoints conversion under: keras-hub/tools/checkpoint_conversion

I don't think this is necessary. Currently esm mainly relies on hf, we just need to provide the hf conversion script

pass-lin commented on keras_hub/src/models/esm/esm_classifier_preprocessor.py:
````diff
@@ -0,0 +1,134 @@
+    Mapping with `tf.data.Dataset`.
+    ```python
+    preprocessor = keras_hub.models.ProteinClassifierPreprocessor.from_preset(
+        "roformer_v2_base_zh"
````
> change the preset name, it's wrong

Can the keras team add presets? Below are my weights, roformer is a native keras2 model, no conversion script is required
https://www.modelscope.cn/profile/q935499957

sachinprasadhs replied:
What I meant is, add preset name for ESM model, not from other models.

pass-lin replied:
> What I meant is, add preset name for ESM model, not from other models.

ok,I have change the preset name

sachinprasadhs commented on keras_hub/src/models/esm/esm_classifier.py:
```diff
@@ -0,0 +1,186 @@
+
+    # Pretrained classifier.
+    classifier = keras_hub.models.ESMProteinClassifier.from_preset(
+        "roformer_v2_base_zh",
```
Write the preset name, like you're mapping in [convert_esm_checkpoints.py](https://github.com/keras-team/keras-hub/pull/2244/files#diff-5865dc50a9921fc9cbdcff2bab772e84be95e7ee2a85fb0682d637c0218987ad)

pass-lin replied:
> Write the preset name, like you're mapping in [convert_esm_checkpoints.py](https://github.com/keras-team/keras-hub/pull/2244/files#diff-5865dc50a9921fc9cbdcff2bab772e84be95e7ee2a85fb0682d637c0218987ad)

OK，i have renewed preset name

pass-lin commented:
> Thanks, few minor comments.
> 
> Also, need more details specific to Keras 3.6 older version issue.
> 
> Finally in the PR description, add the colab notebook to show end to end working of the model, numerics verification. you can follow the PR description template from the recent PR.

How to add a Colab notebook? Can you give me give a demo?

pass-lin commented:
> > > Thanks, few minor comments.
> > > Also, need more details specific to Keras 3.6 older version issue.
> > > Finally in the PR description, add the colab notebook to show end to end working of the model, numerics verification. you can follow the PR description template from the recent PR.
> > 
> > 
> > How to add a Colab notebook? Can you give me give a demo?
> 
> Adding from one of the recent PR which got merged, you can do something like this
> 
> * DeiT Checkpoint Conversion and Numerics Verification Demo (across multiple backends): [Notebook Link](https://colab.research.google.com/drive/15DIJm-pha9LUSOHMZaYB_Ardl4vQXPmK#scrollTo=JU_C_IBGL_WC)
> * DeiT End-to-End Demo (zero-shot and finetuning): [Notebook Link](https://colab.research.google.com/drive/1eGSRfXrfKknUhp0n1nQ62mIQKhjtY3pq#scrollTo=Wjtz0e6j0mr-)
> * [Here](https://drive.google.com/drive/folders/1ni6io9mMzm0lZLv0jwFUv0Q6x2DZ3MzK?usp=drive_link) are the converted DeiT presets from Hugging Face checkpoints for reference.

Hello, I've already added the Colab demo of tools/checkpoint_conversion/convert_esm_checkpoints.py in the PR description. I think this is enough, and we can refer to BERT for the rest.
Can we merge now?

pass-lin commented:
> Hi, The intention of the notebook is to verify the correctness of the model including, backbone, tasks with the usage details and the expected outcome and to verify the numerics stablity after weights transfer to the Keras architecture, with wither forward pass.

Okay, I've added another notebook, which is a demo for predicting the suitable pH of enzymes using ESM.

sachinprasadhs commented:
You can remove the `esm2_t6_8M` directory, that will be generated using the conversion script you have provided and will be uploaded to Kaggle.

The notebook which you have provided doesn't have predict method, 
take any sample suitable input and display the output with predict.

Also in your conversion script, you have mentioned `atol=1e-3`, what would be the error percentage when the `atol=1e-04` and we need following things in your notebook

- Numerics verification, load the original ESM model and do forward pass, and do the same forward pass to Keras-Hub ESM implementation and compare the numerics layer by layer to show if numerics are matching(preferably to the 1e-4 precision)
- Demonstrating usage of proprocessor, Tokenizer and other functionalities of ESM

I have provided the reference notebooks, please refer those.

You can keep only ESM changes in this PR, you can create a new PR for roformer which also needs checkpoint conversion script, so that we can maintain the latest weight in Kaggle by generating the new weights with the script with any future changes to `Keras Hub` model specific.

pass-lin commented:
> You can remove the `esm2_t6_8M` directory, that will be generated using the conversion script you have provided and will be uploaded to Kaggle.
> 
> The notebook which you have provided doesn't have predict method, take any sample suitable input and display the output with predict.
> 
> Also in your conversion script, you have mentioned `atol=1e-3`, what would be the error percentage when the `atol=1e-04` and we need following things in your notebook
> 
> * Numerics verification, load the original ESM model and do forward pass, and do the same forward pass to Keras-Hub ESM implementation and compare the numerics layer by layer to show if numerics are matching(preferably to the 1e-4 precision)
> * Demonstrating usage of proprocessor, Tokenizer and other functionalities of ESM
> 
> I have provided the reference notebooks, please refer those.
> 
> You can keep only ESM changes in this PR, you can create a new PR for roformer which also needs checkpoint conversion script, so that we can maintain the latest weight in Kaggle by generating the new weights with the script with any future changes to `Keras Hub` model specific.

OK, I have modified the notebook, please check. In addition, roformerV2 does not need to convert scripts, it is a native keras model. I just modified the keras2 api

pass-lin commented:
@sachinprasadhs  plz check my notebook

sachinprasadhs commented:
Hi, Still your notebook does not demonstrate the actual use case example demonstrations like https://huggingface.co/docs/transformers/en/model_doc/esm#transformers.EsmForSequenceClassification.forward.example or https://huggingface.co/docs/transformers/en/model_doc/esm#transformers.EsmForProteinFolding.forward.example or 
https://huggingface.co/docs/transformers/en/model_doc/esm#transformers.EsmForTokenClassification.forward.example, please include it.

pass-lin commented:
> Hi, Still your notebook does not demonstrate the actual use case example demonstrations like https://huggingface.co/docs/transformers/en/model_doc/esm#transformers.EsmForSequenceClassification.forward.example or https://huggingface.co/docs/transformers/en/model_doc/esm#transformers.EsmForProteinFolding.forward.example or https://huggingface.co/docs/transformers/en/model_doc/esm#transformers.EsmForTokenClassification.forward.example, please include it.

We've included a [training demo](https://colab.research.google.com/drive/1ALXd1Bq0sWns7ytWzNbwWNanh0YkzBY4?usp=sharing) for ESM. As for ESMFold, that's another brand new pr. So can you just click and tell me what demo to add? Sorry for the trouble.

divyashreepathihalli commented:
/gemini review

pass-lin commented:
> Any demo with the implementation you have which predicts the actual data or the sample input data and display the output in the existing colab, and remove the folder/directory named **esm2_t6_8M** in your code, rest all it looks good. Thanks for all the work.

I’m not sure what you mean by “delete the esm2_t6_8M directory.”

Looking at the [demo notebook](https://colab.research.google.com/drive/15DIJm-pha9LUSOHMZaYB_Ardl4vQXPmK#scrollTo=JU_C_IBGL_WC), all it does is install the environment, change the OS, and then run:

```bash
python tools/checkpoint_conversion/convert_deit_checkpoints.py --preset deit-base-distilled-patch16-384
```

In [my notebook](https://colab.research.google.com/drive/1IPnkQwxZ16E-CfygknU-z8-sCkitgo4l?usp=sharing#scrollTo=tPp192MZ7yn3) I did exactly the same thing: installed the environment, changed the OS, and then ran

```bash
python tools/checkpoint_conversion/convert_esm_checkpoints.py --preset esm2_t6_8M
```

Could you give a more precise and detailed description of which notebook has the problem and what it is missing compared to the reference notebook?  
In the reference notebook, what exactly shows that the esm2_t6_8M directory should be removed?

Further, in [another notebook](https://colab.research.google.com/drive/1ALXd1Bq0sWns7ytWzNbwWNanh0YkzBY4?usp=sharing#scrollTo=p4yCyR7EmglD) I explicitly provide demonstrations of `predict`, `fit`, and `evaluate`. What exactly is still missing?  
A clear description would be greatly appreciated—thank you for your help!

pass-lin commented:
> /gemini review

Thanks, I fixed some bugs with reference to gemini's review.

pass-lin commented:
> In your code commit, there is the files with checkpoint files generated, we don't keep these files in our github, we upload the checkpoints to kaggle/ Hugging face, and the files will be generated by running the conversion script you have provided, you don't need to provide the converted checkpoints here in your commit <img alt="image" width="307" height="350" src="https://private-user-images.githubusercontent.com/73069040/466195521-a071fb83-b764-4e55-ac0d-423c272fcbc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI1NTc5MzQsIm5iZiI6MTc1MjU1NzYzNCwicGF0aCI6Ii83MzA2OTA0MC80NjYxOTU1MjEtYTA3MWZiODMtYjc2NC00ZTU1LWFjMGQtNDIzYzI3MmZjYmM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE1VDA1MzM1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhMzkyZDhlZGMwOTk3MjMzYzk1YTJmY2VhM2VkODdiMWEyOGE2NTJhMmUyY2U3MmY5NzI3YzZkMGM3MzZkZGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-5yoeqyBaRrKgq0NnkT_m-Uk26ow6UR7pJytDQxy8ng">

I'm so sorry, this is an intermediate product when running the test before. I didn't notice his existence. Thanks very much for your reminder.

pass-lin commented:
@sachinprasadhs How far are we from merging now?

pass-lin commented on keras_hub/src/models/esm/esm_encoder.py:
```diff
@@ -0,0 +1,134 @@
+        )
+        residual = x + attention_output
+
+        x = self.feedforward_norm(self.dropout_layer(residual))
```
> This implementation uses a pre-layernorm architecture. However, the original ESM models from Hugging Face (which the weights are being converted from) use a post-layernorm architecture.

https://github.com/huggingface/transformers/blob/fbeaf96f9e2291c21277ac658a33ea8752728bf3/src/transformers/models/esm/modeling_esm.py#L545
ESM is a prenorm Model。There's nothing wrong with that.

pass-lin commented:
> Thank you for the PR @pass-lin. Looks great already. Added a few NIT comments. Thanks for the train demo, do you have a demo that shows output for classifier and PLM with presets loaded?

Thanks for your review, relevant changes have been made.

pass-lin commented on keras_hub/src/models/esm/esm_backbone.py:
```diff
@@ -0,0 +1,230 @@
+    model(input_data)
+
+    # Randomly initialized ESM2 encoder with a custom config.
+    model = keras_hub.models.ESM2Backbone(
+        vocabulary_size=30552,
+        num_layers=4,
+        num_heads=4,
+        hidden_dim=256,
+        intermediate_dim=512,
+        num_heads = 4,
```
> There is a redundant argument again, could you please modify this and add this example in the working demo in the colab

sorry，i have fixed it

pass-lin commented:
<img width="1588" height="346" alt="image" src="https://github.com/user-attachments/assets/83267c66-c9b4-4989-8e4a-59c77ca62358" />
I forgot to switch branches when I was developing other tasks, which led to this error. It has been fixed now
@sachinprasadhs

pass-lin commented:
@sachinprasadhs This test error should have nothing to do with me

pass-lin commented:
> Yes, you can ignore the GPU Test failures

So, can we merge now?

divyashreepathihalli commented:
The GPU tests are fixed, please rebase with master. Will run the GPU tests again

pass-lin commented:
> The GPU tests are fixed, please rebase with master. Will run the GPU tests again

It seems that there are no problems with the test

sachinprasadhs reviewed (approved):
LGTM

pass-lin commented:
@divyashreepathihalli Can this pr merge now?
